### PR TITLE
frontend-forms: Make the validation function Behavior -> BehaviorValidation --- blocked on #49

### PR DESCRIPTION
**Pull request by Ericson2314**
This allows validation to take external state into account. Validation can pull external state, but never be pushed by it, as the validations make clear.

### CODE REVIEW

File: frontend/Rhyolite/Frontend/Form.hs [hunk 1]
```diff
@@ -83,49 +111,68 @@ tagPromptlyDynValidation (DynValidation (Compose b)) = attachPromptlyDynWithMayb
...
+        let msg = errorText typedMsg
         setCustomValidity rawEl msg
         reportValidity_ rawEl
       _ -> setCustomValidity rawEl ("" :: Text) -- NOTE setting empty text is how the browser "clears" the error
   return v
 
+-- | The returned dynamic is really an abuse of I dynamic, namely it shouldn't
+-- be allowed to push anything. The one benefit of this vs 'fmap f . current' is
+-- the result can be fed into 'attachPromptly' rather than 'tag' to avoid
+-- loosing a frame.
```
Comment (3noch) [inline frontend/Rhyolite/Frontend/Form.hs]: `losing`

File: frontend/Rhyolite/Frontend/Form.hs [hunk 2]
```diff
@@ -83,49 +111,68 @@ tagPromptlyDynValidation (DynValidation (Compose b)) = attachPromptlyDynWithMayb
...
     performEvent_ $ ffor (tagPromptlyDyn (fromDynValidation val) validate') $ \case
-      Left msg -> do
+      Left typedMsg -> do
+        let msg = errorText typedMsg
         setCustomValidity rawEl msg
         reportValidity_ rawEl
       _ -> setCustomValidity rawEl ("" :: Text) -- NOTE setting empty text is how the browser "clears" the error
   return v
 
+-- | The returned dynamic is really an abuse of I dynamic, namely it shouldn't
```
Comment (3noch) [inline frontend/Rhyolite/Frontend/Form.hs]: `I dynamic`?

File: frontend/Rhyolite/Frontend/Form.hs [hunk 3]
```diff
@@ -83,49 +111,68 @@ tagPromptlyDynValidation (DynValidation (Compose b)) = attachPromptlyDynWithMayb
...
+-- be allowed to push anything. The one benefit of this vs 'fmap f . current' is
+-- the result can be fed into 'attachPromptly' rather than 'tag' to avoid
+-- loosing a frame.
+weirdMap
```
Comment (3noch) [inline frontend/Rhyolite/Frontend/Form.hs]: This name... :joy: But surely there is a better name...

**Comment by cgibbard** [comment]
Wait, what? Why are we using Behavior now? This is in the context of Reflex-DOM, we might want to show/hide widgets based on the changing validation results, so don't they actually need to be Dynamics?

File: frontend/Rhyolite/Frontend/Form.hs (hunk 3 above)
Comment (Ericson2314) [reply frontend/Rhyolite/Frontend/Form.hs]: good thing it's gone!

**Pull request merged**